src/trajectory/util.cpp

```diff
@@ -0,0 +1,320 @@
+            Eigen::Vector2d{0., segmentDuration}, 4, dimension);
+        problem.addConstantConstraint(0, 0, zeroPos);
+        problem.addConstantConstraint(0, 1, segStartVel);
+        problem.addConstantConstraint(1, 0, segEndPos - segStartPos);
+        problem.addConstantConstraint(1, 1, segEndVel);
+        const auto solution = problem.fit();
+        const auto coefficients = solution.getCoefficients().front();
+        outputTrajectory->addSegment(
+            coefficients, segmentDuration, segmentStartState);
+      }
+      break;
+    }
+
+    currSegmentIdx++;
+    currSegmentStartTime = currSegmentEndTime;
+  }
+
+  for (std::size_t i = currSegmentIdx + 1; i < traj.getNumSegments(); i++)
+  {
+    outputTrajectory->addSegment(
+        traj.getSegmentCoefficients(i),
+        traj.getSegmentDuration(i),
+        traj.getSegmentState(i));
+  }
+  return outputTrajectory;
+}
+
+//==============================================================================
+aikido::trajectory::UniqueInterpolatedPtr convertToInterpolated(
+    const aikido::trajectory::Spline& traj,
+    aikido::statespace::ConstInterpolatorPtr& interpolator)
+{
+  auto stateSpace = traj.getStateSpace();
+  auto outputTrajectory
+      = make_unique<aikido::trajectory::Interpolated>(stateSpace, interpolator);
+
+  auto state = stateSpace->createState();
+  double t = 0.0;
+  for (std::size_t i = 0; i < traj.getNumWaypoints(); i++)
+  {
+    traj.getWaypoint(i, state);
+    t = traj.getWaypointTime(i);
+    outputTrajectory->addWaypoint(t, state);
+  }
+
+  return outputTrajectory;
+}
+
+//==============================================================================
+aikido::trajectory::UniqueInterpolatedPtr concatenate(
+    const aikido::trajectory::Interpolated& traj1,
+    const aikido::trajectory::Interpolated& traj2)
+{
+  if (traj1.getStateSpace()->getDimension()
+      != traj2.getStateSpace()->getDimension())
+    throw std::runtime_error("Dimension mismatch");
+
+  auto stateSpace = traj1.getStateSpace();
+
+  auto outputTrajectory = make_unique<aikido::trajectory::Interpolated>(
+      stateSpace, traj1.getInterpolator());
+
+  for (std::size_t i = 0; i < traj1.getNumWaypoints() - 1; i++)
+  {
+    outputTrajectory->addWaypoint(
+        traj1.getWaypointTime(i), traj1.getWaypoint(i));
+  }
+  outputTrajectory->addWaypoint(traj1.getEndTime(), traj2.getWaypoint(0));
+  double endTimeOfTraj1 = traj1.getEndTime();
+  for (std::size_t i = 1; i < traj2.getNumWaypoints(); i++)
+  {
+    outputTrajectory->addWaypoint(
+        endTimeOfTraj1 + traj2.getWaypointTime(i), traj2.getWaypoint(i));
+  }
+  return outputTrajectory;
+}
+
+//==============================================================================
+aikido::trajectory::UniqueSplinePtr concatenate(
+    const aikido::trajectory::Spline& traj1,
+    const aikido::trajectory::Spline& traj2)
+{
+  if (traj1.getStateSpace() != traj2.getStateSpace())
+    throw std::runtime_error("State space mismatch");
```

Comment:
`concatenate()` for `Interpolated` just checks the dimension while this function checks the instances.

@gilwoolee @brianhou Hm, what was the correct way of checking the compatibility of state spaces?

Reply:
Since we're merging the two trajectories in this case, shouldn't this check for instance match?